File: simple-castle-project/core/src/main/java/com/simple/castle/utils/ModelUtils.java
package com.simple.castle.utils;

import com.badlogic.gdx.graphics.g3d.Model;

import java.util.Collection;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public final class ModelUtils {
    private ModelUtils() {

    }

    public static Collection<String> getNodesFromModelByPattern(Model model, String pattern) {
        return getValuesByPattern(
                StreamSupport.stream(model.nodes.spliterator(), false)
                        .map(node -> node.id)
                        .collect(Collectors.toSet()), pattern);
    }

    public static Collection<String> getValuesByPattern(Set<String> modelNames, String pattern) {
        return modelNames.stream()
                .filter(nodes -> nodes.matches(pattern))
                .collect(Collectors.toList());
    }
}

File: core/common/src/main/kotlin/castle/core/util/ModelUtils.kt
package castle.core.util

import com.badlogic.gdx.graphics.g3d.Model
import com.badlogic.gdx.math.Vector3

interface ModelUtils {
    companion object {
        val DEFAULT_FACE_DIRECTION: Vector3 = Vector3(1f, 0f, 0f)

        fun searchNodes(models: Map<String, Model>, nodesPattern: String): Collection<String> {
            return models
                .flatMap { it.value.nodes }
                .map { node -> node.id }
                .toSet()
                .filter { nodes -> nodes.matches(Regex(nodesPattern)) }
        }
    }
}